### DISCUSSION TITLE
Avalonia window (textboxes) embedded in Autodesk Inventor not accepting input

### DISCUSSION THREAD

**Discussion by Basnederveen** [AvaloniaUI/Avalonia#13731] [Q&A] [answered]
See my post on StackOverflow: https://stackoverflow.com/questions/77543411/avalonia-window-textboxes-embedded-in-autodesk-inventor-not-accepting-input

Creating an Autodesk Inventor plugin. Have a small avalonia app which I want to host in the dockable window in Inventor.

https://stackoverflow.com/questions/835878/wpf-textbox-not-accepting-input-when-in-elementhost-in-window-forms

Similar problem to above. There is something I'm missing about how to forward the keystrokes to the avalonia component. 

The key events are fired in the MainWindow code-behind, but they are not forwarded to the AvaloniaWpfHost control.

**Comment by MrJul** [comment, marked as answer]
I'm not sure why you're creating a WPF window as an intermediate if you only want Avalonia controls? Thatintermediate layer doesn't seem necessary. As far as I can tell, Inventor seems to only need a standard Win32 HWND.

You can create an Avalonia `Window` and get its HWND through `Window.TryGetPlatformHandle().Handle`, that you should be able to pass directly to Inventor's `DockableWindow.AddChild()`.

Then, to handle the `WM_GETDLGCODE` message, you have two choices:
 - You're willing to use latest 11.1 nightly builds: use `Win32SpecificOptions.AddWndProcHookCallback()`
 - You're on stable 11.0.x: use the standard Win32 [`SetWindowLongPtr`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowlongptrw) function with `GWLP_WNDPROC` to change the window procedure.

**Comment by Basnederveen** [comment]
Hi MrJul,

Thank you so much for the reply. I've just created the WPF window because I hoped it would fix the problem. I can indeed add the avalonia window to the DockableWindow by its handle.

Since I'm on 11.0.x I tried the SetWindowLongPtr function. But this crashes Inventor..

I'm calling the function straight after adding the child to the dockable window.

**Reply by MrJul** [reply]
That's probably a bad usage of `SetWindowLongPtr`, like wrong argument types defined for your WndProc callback, corrupting the stack. Also make sure to call the old window proc for messages others than `WM_GETDLGCODE`, or your window won't work correctly.

If you're not comfortable with P/Invoke, I'd suggest to try the nightly builds since you'll get the nice `AddWndProcHookCallback` instead of having to replace the WndProc manually. This way you'll be able to validate whether handling the message solves the problem in Inventor. Then you can figure the P/Invoke stuff to port that to 11.0.

**Reply by Basnederveen** [reply]
That's probably likely. This is all quite new for me so I thought you meant I should just call the SetWindowLongPtr in my initialization code.

This crashes. I see now you wrote GWLP_WNDPROC not GWL_WNDPROC. I thought at first you meant to use the function as it is used in the SetParent function of https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/WindowImpl.cs#L656

With the WindowLongParam GWL_WNDPROC (-4). 


    private void InitializePropertyEditor()
    {
        propertyWindow = TryGetDockableWindow();
        services = Builder.BuildLauncherServices();
        builder = Builder.BuildAvaloniaAppWithServices(services).SetupWithoutStarting();

        MainWindow = new EmbeddedWindow()
        {
            DataContext = services.GetService<MainViewModel>()
        };

        MainWindow.Show();

        var hwnd = MainWindow.TryGetPlatformHandle().Handle;
        propertyWindow.AddChild(hwnd);

        //SetWindowLongPtr(hwnd, -4, propertyWindow.HWND);
    }


I'm at the moment trying the nightly build. But the latest one I see in the nuget source is 11.0.999. It contains the AddWndProcHookCallback, so I guess this is the one you mean.

**Reply by MrJul** [reply]
Note: 11.0.999 is a "wrong" release (well for your needs, it already works), almost 2 months old. You have to check the _Pre-release_ checkbox when adding a NuGet package and use the latest one, whose version should be something like `11.1.999-cibuild00XXXXX-beta`.

**Reply by Basnederveen** [reply]
Got it, thanks!

**Comment by Basnederveen** [comment]
It works!!!! Thanks again!! Had a feeling it would be a one-line fix..
       

    private void InitializePropertyEditor()
    {
        propertyWindow = TryGetDockableWindow();
        var services = Builder.BuildLauncherServices();
        var builder = Builder.BuildAvaloniaAppWithServices(services).SetupWithoutStarting();

        MainWindow = new EmbeddedWindow()
        {
            DataContext = services.GetService<MainViewModel>()
        };

        MainWindow.Show();

        var hwnd = MainWindow.TryGetPlatformHandle().Handle;
        propertyWindow.AddChild(hwnd);
        
        Win32SpecificOptions.AddWndProcHookCallback(TopLevel.GetTopLevel(MainWindow), WndProc);
    }

    private DockableWindow TryGetDockableWindow()
    {
        var internalName = "BasAutomation.PropertyEditor";
        var windowName = "Properties";

        DockableWindow propertyWindow;
        try
        {
            propertyWindow = app.UserInterfaceManager.DockableWindows[internalName];
        }
        catch (Exception ex)
        {
            propertyWindow = app.UserInterfaceManager.DockableWindows.Add(AddInClientId, internalName, windowName);
        }

        return propertyWindow;
    }

    private const UInt32 DLGC_WANTARROWS = 0x0001;
    private const UInt32 DLGC_WANTTAB = 0x0002;
    private const UInt32 DLGC_WANTALLKEYS = 0x0004;
    private const UInt32 DLGC_HASSETSEL = 0x0008;
    private const UInt32 DLGC_WANTCHARS = 0x0080;
    private const UInt32 WM_GETDLGCODE = 0x0087;

    private static IntPtr WndProc(IntPtr hWnd, uint msg, IntPtr wParam, IntPtr lParam, ref bool handled)
    {
        if (msg == WM_GETDLGCODE)
        {
            handled = true;
            return new IntPtr(DLGC_WANTCHARS | DLGC_WANTARROWS | DLGC_HASSETSEL | DLGC_WANTTAB | DLGC_WANTALLKEYS);
        }

        return IntPtr.Zero;
    }
}

**Reply by MrJul** [reply]
You're welcome. Happy you could get it working!
